README.md
weewee
======

Wrapper Extension Engine for veewee

A ruby script for generating veewee templates based on some defaults

Currently has basic functionality for Linux. Planning to add Solaris.

Usage
=====

	$ weewee.rb -[DOdjklpsfvn:o:p:r:L]

	-V: Display version information
	-h: Display usage information
	-k: Create kickstart file (Also sets OS type to Linux)
	-l: Set OS to Linux
	-s: Set OS to Solaris
	-r: Remove definition for host
	-n: Name of new machine to be created
	-p: Create postinstall file
	-d: Create definitions file
	-D: Use defaults
	-v: Verbose output
	-f: Output to file rather than STDOUT
	-L: List Veewee OS types
	-T: List Veewee Templates
	-p: Set OS type
	-O: Set virtualisation platform to Virtual Box
	-F: Set virtualisation platform to VMware Fusion

Examples
========

Generate Linux Kickstart file and review defaults:

	$ weewee.rb -l -k -n HOSTNAME
	OS Type is Linux
	Loading ./methods/definition.rb
	Loading ./methods/kickstart.rb
	Loading ./methods/postinstall.rb

	You will be presented with a set of questions followed by the default output
	If you are happy with the default output simply hit enter

	Install type?
	[ install ]
  
	Install Medium?
	[ cdrom ]
  
	Install Language?
	[ en_US.UTF-8 ]

Delete a definition:

	$ veewee.rb -r HOSTNAME

Generate Linux Kickstart file to STDOUT and accept defaults:

	$ weewee.rb -l -k -D -n HOSTNAME
	OS Type is Linux
	Loading ./methods/definition.rb
	Loading ./methods/kickstart.rb
	Loading ./methods/postinstall.rb
	# weewee (Wrapper Extension Engine for veewee) v. 0.0.1 Richard Spindler <richard@lateralblast.com.au>
	install
	cdrom
	lang en_US.UTF-8
	langsupport --default=en_US.UTF-8 en_US.UTF-8
	keyboard us
	rootpw --iscrypted $1Eos8GLxaqSA
	network --device eth0 --bootproto dhcp
	selinux --enforcing
	authconfig --enableshadow --enablemd5
	timezone Australia/Melbourne
	bootloader --location=mbr
	clearpart --all --drives=sda --initlabel
	part /boot --fstype ext3 --size=100 --ondisk=sda
	part pv.2 --size=0 --grow --ondisk=sda
	volgroup VolGroup00 --pesize=32768 pv.2
	logvol swap --fstype swap --name=LogVol01 --vgname=VolGroup00 --size=512 --grow --maxsize=1024
	logvol / --fstype ext3 --name=LogVol00 --vgname=VolGroup00 --size=1024 --grow
	reboot

	%packages
	@ core
	grub
	e2fsprogs
	lvm2
	kernel-devel
	kernel-headers

	%post
	/usr/sbin/groupadd vagrant
	/usr/sbin/useradd vagrant -g vagrant -G wheel
	echo "vagrant"|passwd --stdin vagrant
	echo "vagrant        ALL=(ALL)       NOPASSWD: ALL" >> /etc/sudoers

Generate Linux Kickstart file ks.cfg  and accept defaults:
	
	$ weewee.rb -l -k -D -f -n HOSTNAME

Generate Definitions file to STDOUT and accept defaults:

	$ weewee.rb -l -d -D -n HOSTNAME
	OS Type is Linux
	Loading ./methods/definition.rb
	Loading ./methods/kickstart.rb
	Loading ./methods/postinstall.rb
	Veewee::Session.declare({
	  :cpu_count => "1",
	  :memory_size => "384",
	  :disk_size => "10240",
	  :disk_format => "VDI",
	  :hostiocache => "off",
	  :ioapic => "on",
	  :pae => "on",
	  :os_type_id => "RedHat_64",
	  :boot_wait => "10",
	  :iso_src => "CentOS-5.9-x86_64-bin-DVD-1of2.iso",
	  :iso_md5 => "",
	  :iso_download_timeout => "10000",
	  :boot_cmd_sequence => "[ 'linux text ks=http://%IP%:%PORT%/ks.cfg<Enter>' ]",
	  :kickstart_port => "7122",
	  :kickstart_timeout => "10000",
	  :kickstart_file => "ks.cfg",
	  :ssh_login_timeout => "10000",
	  :ssh_user => "vagrant",
	  :ssh_password => "vagrant",
	  :ssh_key => "",
	  :ssh_host_port => "7222",
	  :ssh_guest_port => "22",
	  :shutdown_cmd => "/sbin/halt -h -p",
	  :postinstall_files => "[ 'postinstall.sh' ]",
	  :postinstall_timeout => "10000"
	})



changelog
# Changes:
# 0.0.1 Thu  8 Aug 2013 22:39:21 EST
# Initial commit
# 0.0.2 Fri  9 Aug 2013 11:01:05 EST
# Fixed issue with definitions output
# 0.0.3 Sat 10 Aug 2013 16:25:22 EST
# Cleaned up code and added initial Solaris support
# 0.0.4 Sun 11 Aug 2013 18:29:26 EST
# Improved Linux and Solaris support
# 0.0.5 Thu 31 Oct 2013 11:19:23 EST
# Cleaned up code and added capability to delete definitions
# 0.0.6 Thu 31 Oct 2013 11:31:49 EST
# Fixed command line arguement handling
# 0.0.7 Thu 31 Oct 2013 15:52:29 EST
# Added support for using veewee defaults


weewee.rb
#!/usr/bin/env ruby

# Name:         weewee (Wrapper Extension Engine for veewee)
# Version:      0.0.7
# Release:      1
# License:      Open Source
# Group:        System
# Source:       N/A
# URL:          http://lateralblast.com.au/
# Distribution: UNIX
# Vendor:       Lateral Blast
# Packager:     Richard Spindler <richard@lateralblast.com.au>
# Description:  Ruby script to wrapper veewee

require 'getopt/std'
require 'fileutils'
require 'veewee'

# Set some defaults

default_platform="fusion"
default_linux="CentOS-5.9-x86_64"
default_solaris="solaris-10-ga-x86"
$verbose=1
$iso_store=""
$base_dir=""

# Create a structure to store kickstart/jumpstart information
# Type determines whether it's output to the config or is used by another object

Cfg=Struct.new(:type, :question, :parameter, :value, :eval)
Dfn=Struct.new(:question, :value)
Pkg=Struct.new(:name, :install)

def get_version()
  file_array=IO.readlines $0
  version=file_array.grep(/^# Version/)[0].split(":")[1].gsub(/^\s+/,'').chomp
  packager=file_array.grep(/^# Packager/)[0].split(":")[1].gsub(/^\s+/,'').chomp
  name=file_array.grep(/^# Name/)[0].split(":")[1].gsub(/^\s+/,'').chomp
  version=name+" v. "+version+" "+packager
  return version
end

def print_version()
  version=get_version()
  puts version
end

def print_usage(options)
  puts
  print_version()
  puts
  puts "Usage: "+$0+" -["+options+"]"
  puts
  puts "-V: Display version information"
  puts "-h: Display usage information"
  puts "-k: Create kickstart file (Also sets OS type to Linux)"
  puts "-W: Don't apply weewee modifications"
  puts "-l: Set OS to Linux"
  puts "-s: Set OS to Solaris"
  puts "-r: Remove definition for host"
  puts "-n: Name of new machine to be created"
  puts "-p: Create postinstall file"
  puts "-d: Create definitions file"
  puts "-D: Use defaults"
  puts "-v: Verbose output"
  puts "-f: Output to file rather than STDOUT"
  puts "-L: List Veewee OS types"
  puts "-T: List Veewee Templates"
  puts "-p: Set OS type"
  puts "-O: Set virtualisation platform to Virtual Box"
  puts "-F: Set virtualisation platform to VMware Fusion"
  puts
end

def get_password_crypt(password)
  possible=[('a'..'z'),('A'..'Z'),(0..9),'.','/'].inject([]) {|s,r| s+Array(r)}
  salt=Array.new(8){possible[rand(possible.size)]}
  password=password.crypt("$1$#{salt}")
  return password
end

options="DLOWdjklpsfvn:o:p:r:"

# Get command line arguments
# Print help if given none

if !ARGV[0]
  print_usage(options)
  exit
end

begin
  opt=Getopt::Std.getopts(options)
rescue
  print_usage(options)
  exit
end

if opt["h"]
  print_usage(options)
  exit
end

# Check local config

def check_local_config()
  iso_dir=Dir.home+"/ISOs"
  if Dir.exists?(iso_dir)
    $iso_store=iso_dir+"/"
    if $verbose == 1
      puts "Setting ISO directory to "+$iso_store
    end
  end
end

check_local_config()

# Set up base directory

def create_dir(dir_name)
  if !Dir.exists?(dir_name)
    if $verbose == 1
      puts "Creating directory: "+dir_name
    end
    Dir.mkdir(dir_name)
  end
end

base_dir=Dir.home+"/.weewee"
create_dir(base_dir)

def delete_config(config_name,base_dir)
  definition_dir=base_dir+"/definitions/"+config_name
  if Dir.exists?(definition_dir)
    puts "Deleting "+definition_dir 
    FileUtils.rm_r(definition_dir)
  else
    puts "veewee definition for "+config_name+" does not exist in directory "+definition_dir
  end
  return
end

if opt["r"]
  config_name=opt["r"]
  delete_config(config_name,base_dir)
  exit
end

if opt["F"] or opt ["O"]
  if opt ["F"]
    platform="fusion"
  else
    if opt ["O"]
      platform="vbox"
    end
  end
else
  platform=default_platform
end

if $verbose == 1
  puts "Setting virtualisation platform to "+platform
end

if opt["L"]
  vcli=Veewee::Environment.new
  vcli.ostypes.each do |key, value|
    if value[:vbox]
      puts value[:vbox]
    end
  end
  exit
end

if opt["T"]
  vcli=Veewee::Environment.new
  template = Veewee::Templates.new(vcli)
  template.each do |key, value|
    puts key
  end
  exit
end

if opt["v"]
  $verbose=1
else
  $verbose=0
end

# If given -l set OS type to Linux

if opt["l"]
  os_type="linux"
  template_name=default_linux
  if opt["v"]
    puts "OS type set to Linux"
  end
end

# If given -s set OS type to Solaris

if opt["s"]
  os_type="solaris"
  template_name=default_solaris
  if opt["v"]
    puts "OS type set to Solaris"
  end
end

if opt["n"]
  if !opt["n"]
    puts "Machine name not specified"
    exit
  end
  box_name=opt["n"]
  system("cd #{base_dir} ; veewee #{platform} define '#{box_name}' '#{template_name}' ")
end

base_dir=base_dir+"/definitions"

create_dir(base_dir)

if opt["f"]
  create_file=1
end

# If OS type is not set handle appropriately
# If -k set to Linux as it's Kickstart
# If -j set to Solaris as it's Jumpstart

if !opt["s"] and !opt["l"]
  if opt["k"]
    os_type="linux"
    puts "Setting OS type to Linux"
  else 
    if opt["j"]
      os_type="solaris"
      puts "Setting OS type to Solaris"
    else
      puts "OS type not selected"
      exit
    end
  end
end

if opt["W"]
  if $verbose == 1
    puts "Using veewee defaults"
  end
end

# If given -k and no -s (ie not Solaris) build kickstart file

if !opt["W"]
  # If given any option that requires methods, load them
  if opt["k"] or opt["p"] or opt["d"] or opt["j"] or opt["s"]  or opt["l"]
    # Import methods
    if Dir.exists?("./methods")
      file_list=Dir.entries("./methods")
      for file in file_list
        if file =~/rb$/
          if $verbose == 1
            puts "Loading ./methods/"+file
          end
          require "./methods/#{file}"
        end
      end
    end
  end
  if opt["D"]
    output_type="default"
    if $verbose == 1
      puts "Using defaults"
    end
  end
  if opt["k"] or opt["l"] and !opt["s"]
    create_file=opt["f"]
    config_name=opt["n"]
    if $verbose == 1
      puts "Generating Kickstart file"
    end
    do_kickstart(config_name,create_file,output_type,base_dir)
  end
  if opt["p"] or opt["s"] or opt["l"]
    create_file=opt["f"]
    config_name=opt["n"]
    if !opt["l"] and !opt["s"]
      puts "OS type not specified"
      exit
    end
    if $verbose == 1
      puts "Generating Postinstall file"
    end
    do_postinstall(config_name,create_file,output_type,os_type,base_dir,platform)
  end
  if opt["j"] or opt["s"] and !opt["l"]
    create_file=opt["f"]
    config_name=opt["n"]
    if $verbose == 1
      puts "Generating Jumpstart file"
    end
    do_jumpstart(config_name,create_file,output_type,base_dir)
  end
  if opt["d"]  or opt["l"] or opt["s"]
    create_file=opt["f"]
    config_name=opt["n"]
    if !opt["l"] and !opt["s"]
      puts "OS type not specified"
      exit
    end
    if $verbose == 1
      puts "Generating Definition file"
    end
    do_definition(config_name,create_file,output_type,os_type,base_dir)
  end
end


methods/definition.rb

def populate_postinstall_definition(str,os_type)
  cfg=Dfn.new(
    question  = "Postinstall Files",
    value     = "[ 'postinstall.sh' ]"
    )
  str["postinstall_files"]=cfg
  cfg=Dfn.new(
    question  = "Postinstall timeout",
    value     = "10000"
    )
  str["postinstall_timeout"]=cfg
  return str
end

def populate_shutdown_definition(str,os_type)
  cfg=Dfn.new(
    question  = "Shutdown command",
    value     = "/sbin/halt -h -p"
    )
  str["shutdown_cmd"]=cfg
  return str
end

def populate_sudo_definition(str,os_type)
  return str
  cfg=Dfn.new(
    question  = "Sudo command",
    value     = "echo '%p'|sudo -S sh '%f'"
    )
  str["sudo_cmd"]=cfg
  return str
end

def populate_system_definition(str,os_type)
  cfg=Dfn.new(
    question  = "Number of CPUs",
    value     = "1"
    )
  str["cpu_count"]=cfg
  if os_type == "linux"
    cfg=Dfn.new(
      question  = "Memory size",
      value     = "384"
      )
  end
  if os_type == "solaris"
    cfg=Dfn.new(
      question  = "Memory size",
      value     = "768"
      )
  end
  str["memory_size"]=cfg
  if os_type == "linux"
    cfg=Dfn.new(
      question  = "Disk size",
      value     = "10140"
      )
  end
  if os_type == "solaris"
    cfg=Dfn.new(
      question  = "Disk size",
      value     = "65140"
      )
  end
  str["disk_size"]=cfg
  cfg=Dfn.new(
    question  = "Disk format",
    value     = "VDI"
    )
  str["disk_format"]=cfg
  cfg=Dfn.new(
    question  = "Host IO cache",
    value     = "off"
    )
  str["hostiocache"]=cfg
  cfg=Dfn.new(
    question  = "IO APIC",
    value     = "on"
    )
  str["ioapic"]=cfg
  cfg=Dfn.new(
    question  = "PAE",
    value     = "on"
    )
  str["pae"]=cfg
  return str
end

def populate_os_definition(str,os_type)
  if os_type == "linux"
    cfg=Dfn.new(
      question  = "OS type ID",
      value     = "RedHat_64"
      )
  end
  if os_type == "solaris"
    cfg=Dfn.new(
      question  = "OS type ID",
      value     = "Solaris_64"
      )
  end
  str["os_type_id"]=cfg
  return str
end

def populate_boot_definition(str,os_type)
  cfg=Dfn.new(
    question  = "Boot wait",
    value     = "10"
    )
  str["boot_wait"]=cfg
  return str
end

def populate_iso_definition(str,os_type)
  if os_type == "linux"
    cfg=Dfn.new(
      question  = "ISO File",
      value     = "#{$iso_store}CentOS-5.9-x86_64-bin-DVD-1of2.iso"
      )
  end
  if os_type == "solaris"
    cfg=Dfn.new(
      question  = "ISO File",
      value     = "#{$iso_store}sol-10-u11-ga-x86-dvd.iso"
      )
  end
  str["iso_src"]=cfg
  cfg=Dfn.new(
    question  = "ISO MD5",
    value     = ""
    )
  str["iso_md5"]=cfg
  cfg=Dfn.new(
    question  = "ISO download timeout",
    value     = "10000"
    )
  str["iso_download_timeout"]=cfg
  return str
end

def populate_linux_definition(str,os_type)
  if os_type == "linux"
    cfg=Dfn.new(
      question  = "Boot command sequence",
      value     = "[ 'linux text ks=http://%IP%:%PORT%/ks.cfg<Enter>' ]"
      )
  end
  if os_type == "solaris"
    cfg=Dfn.new(
      question  = "Boot command sequence",
      value     = "[ 'e', 'e', '<Backspace>'*22, '- nowin install -B install_media=cdrom<Enter>', 'b', ]" 
      )
  end
  str["boot_cmd_sequence"]=cfg
  if os_type == "solaris"
    cfg=Dfn.new(
      question  = "Floppy files",
      value     = "[ 'jumpstart/sysidcfg', 'jumpstart/rules', 'jumpstart/rules.ok',
                     'jumpstart/begin', 'jumpstart/profile', 'jumpstart/finish' ]"
      )
    str["floppy_file"]=cfg
  end
  cfg=Dfn.new(
    question  = "Kickstart port",
    value     = "7122"
    )
  str["kickstart_port"]=cfg
  cfg=Dfn.new(
    question  = "Kickstart timeout",
    value     = "10000"
    )
  str["kickstart_timeout"]=cfg
  cfg=Dfn.new(
    question  = "Kickstart file",
    value     = "ks.cfg"
    )
  str["kickstart_file"]=cfg
  return str
end

def populate_ssh_definition(str,os_type)
  cfg=Dfn.new(
    question  = "SSH login timeout",
    value     = "10000"
    )
  str["ssh_login_timeout"]=cfg
  cfg=Dfn.new(
    question  = "SSH user",
    value     = "vagrant"
    )
  str["ssh_user"]=cfg
  cfg=Dfn.new(
    question  = "SSH password",
    value     = "vagrant"
    )
  str["ssh_password"]=cfg
  cfg=Dfn.new(
    question  = "SSH Key",
    value     = ""
    )
  str["ssh_key"]=cfg
  cfg=Dfn.new(
    question  = "SSH Host port",
    value     = "7222"
    )
  str["ssh_host_port"]=cfg
  cfg=Dfn.new(
    question  = "SSH guest port",
    value     = "22"
    )
  str["ssh_guest_port"]=cfg
  return str
end

def output_definition(str,output_file)
  output="Veewee::Session.declare({\n"
  if output_file
    file=File.open(output_file, 'a')
    file.write(output)
  else
    print output 
  end 
  str.each do |key, value|
    if key == "postinstall_timeout"
      output="  :"+key+" => \""+str[key].value+"\"\n"
    else
      output="  :"+key+" => \""+str[key].value+"\",\n"
    end
    if output_file
      file.write(output)
    else
      print output 
    end
  end
  if output_file
    output="})"
    file.write(output)
    file.close
  else
    puts "})"
  end
  return
end

def verify_definition(str,os_type)
  puts
  puts "You will be presented with a set of questions followed by the default output"
  puts "If you are happy with the default output simply hit enter"
  puts
  str.each do |key, value|
    puts str[key].question+"?"
    puts "[ "+str[key].value+" ]"
    print "  "
    answer=gets.chomp
    if answer != ""
      if answer != str[key].value
        str[key].value=answer 
      end
    end
  end
  return str
end

def do_definition(config_name,output_file,output_type,os_type,base_dir)
  if output_file
    output_file="definition.rb"
    if config_name
      output_dir=base_dir+"/"+config_name
      output_file=output_dir+"/"+output_file
      if !Dir.exists?(output_dir)
        Dir.mkdir(output_dir)
      end
    end
    if File.exists?(output_file)
      File.delete(output_file)
    end
  end
  str=Hash.new
  str=populate_definition(str,os_type)
  if output_type != "default"
    str=verify_definition(str,os_type)
  end
  output_definition(str,output_file)
  if output_file
    FileUtils.chmod(0755,output_file)
  end
end

def populate_definition(str,os_type)
  str=populate_system_definition(str,os_type)
  str=populate_os_definition(str,os_type)
  str=populate_boot_definition(str,os_type)
  str=populate_iso_definition(str,os_type)
  if os_type == "linux"
    str=populate_linux_definition(str,os_type)
  end
  str=populate_ssh_definition(str,os_type)
  str=populate_sudo_definition(str,os_type)
  str=populate_shutdown_definition(str,os_type)
  str=populate_postinstall_definition(str,os_type)
  return str
end


methods/kickstart.rb

def get_kickstart_language(str)
  result="--default="+str["install_language"].value+" "+str["install_language"].value
  return result
end

def get_kickstart_xconfig(str)
  result="--card "+str["videocard"].value+" --videoram "+str["videoram"].value+" --hsync "+str["hsync"].value+" --vsync "+str["vsync"].value+" --resolution "+str["resolution"].value+" --depth "+str["depth"].value
  return result
end

def get_kickstart_network(str)
  result="--device "+str["nic"].value+" --bootproto "+str["bootproto"].value
  return result
end

def get_kickstart_password(str)
  result="--iscrypted "+str["crypt"].value
  return result
end

def get_kickstart_bootloader(str)
  result="--location="+str["bootstrap"].value
  return result
end

def get_kickstart_clearpart(str)
  result="--all --drives="+str["bootdevice"].value+" --initlabel"
  return result
end

def get_kickstart_bootpart(str)
  result=str["bootmount"].value+" --fstype "+str["bootfs"].value+" --size="+str["bootsize"].value+" --ondisk="+str["bootdevice"].value
  return result
end

def get_kickstart_volpart(str)
  result=str["volname"].value+" --size="+str["volsize"].value+" --grow --ondisk="+str["bootdevice"].value
  return result
end

def get_kickstart_volgroup(str)
  result=str["volgroupname"].value+" --pesize="+str["pesize"].value+" "+str["volname"].value
  return result
end

def get_kickstart_logswap(str)
  result="swap --fstype swap --name="+str["swapvol"].value+" --vgname="+str["volgroupname"].value+" --size="+str["swapmin"].value+" --grow --maxsize="+str["swapmax"].value
  return result
end

def get_kickstart_logroot(str)
  result="/ --fstype "+str["rootfs"].value+" --name="+str["rootvol"].value+" --vgname="+str["volgroupname"].value+" --size="+str["rootsize"].value+" --grow"
  return result
end

def populate_kickstart_header(str)
  version=get_version()
  cfg=Cfg.new(
    type      = "output",
    question  = "Kickstart file header comment",
    parameter = "",
    value     = "# "+version,
    eval      = "no"
    )
  str["kickstart_header"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Install type",
    parameter = "",
    value     = "install",
    eval      = "no"
    )
  str["install_type"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Install Medium",
    parameter = "",
    value     = "cdrom",
    eval      = "no"
    )
  str["install_method"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Install Language",
    parameter = "lang",
    value     = "en_US.UTF-8",
    eval      = "no"
    )
  str["install_language"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Support Language",
    parameter = "langsupport",
    value     = get_kickstart_language(str),
    eval      = "get_kickstart_language(str)"
    )
  str["support_language"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Keyboard",
    parameter = "keyboard",
    value     = "us",
    eval      = "no"
    )
  str["keyboard"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Video Card",
    parameter = "",
    value     = "VMWare",
    eval      = "no"
    )
  str["videocard"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Video RAM",
    parameter = "",
    value     = "16384",
    eval      = "no"
    )
  str["videoram"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Horizontal Sync",
    parameter = "",
    value     = "31.5-37.9",
    eval      = "no"
    )
  str["hsync"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Vertical Sync",
    parameter = "",
    value     = "50-70",
    eval      = "no"
    )
  str["vsync"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Resolution",
    parameter = "",
    value     = "800x600",
    eval      = "no"
    )
  str["resolution"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Bit Depth",
    parameter = "",
    value     = "16",
    eval      = "no"
    )
  str["depth"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Xconfig",
    parameter = "xconfig",
    value     = get_kickstart_xconfig(str),
    eval      = "get_kickstart_xconfig(str)"
    )
  str["xconfig"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Primary Network Interface",
    parameter = "",
    value     = "eth0",
    eval      = "no"
    )
  str["nic"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Boot Protocol",
    parameter = "",
    value     = "dhcp",
    eval      = "no"
    )
  str["bootproto"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Network Configuration",
    parameter = "network",
    value     = get_kickstart_network(str),
    eval      = "get_kickstart_network(str)"
    )
  str["network"]=cfg
  password="sun123"
  cfg=Cfg.new(
    type      = "",
    question  = "Root Password",
    parameter = "",
    value     = password,
    eval      = "no"
    )
  str["password"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Root Password Crypt",
    parameter = "",
    value     = get_password_crypt(password),
    eval      = "get_password_crypt(password)"
    )
  str["crypt"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Root Password Configuration",
    parameter = "rootpw",
    value     = get_kickstart_password(str),
    eval      = "get_kickstart_password(str)"
    )
  str["bootproto"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Root Password Configuration",
    parameter = "rootpw",
    value     = get_kickstart_password(str),
    eval      = "get_kickstart_password(str)"
    )
  str["bootproto"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "SELinux Configuration",
    parameter = "selinux",
    value     = "--enforcing",
    eval      = "no"
    )
  str["selinux"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "SELinux Configuration",
    parameter = "authconfig",
    value     = "--enableshadow --enablemd5",
    eval      = "no"
    )
  str["authconfig"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Timezone",
    parameter = "timezone",
    value     = "Australia/Melbourne",
    eval      = "no"
    )
  str["timezone"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Bootstrap",
    parameter = "",
    value     = "mbr",
    eval      = "no"
    )
  str["bootstrap"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Bootloader",
    parameter = "bootloader",
    value     = get_kickstart_bootloader(str), 
    eval      = "get_kickstart_bootloader(str)"
    )
  str["bootloader"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Boot Device",
    parameter = "",
    value     = "sda",
    eval      = "no"
    )
  str["bootdevice"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Clear Parition",
    parameter = "clearpart",
    value     = get_kickstart_clearpart(str), 
    eval      = "get_kickstart_clearpart(str)"
    )
  str["clearpart"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Boot Mount Point",
    parameter = "",
    value     = "/boot",
    eval      = "no"
    )
  str["bootmount"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Boot Filesystem",
    parameter = "",
    value     = "ext3",
    eval      = "no"
    )
  str["bootfs"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Boot Size",
    parameter = "",
    value     = "100",
    eval      = "no"
    )
  str["bootsize"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Clear Parition",
    parameter = "part",
    value     = get_kickstart_bootpart(str), 
    eval      = "get_kickstart_bootpart(str)"
    )
  str["bootpart"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Physical Volume Name",
    parameter = "",
    value     = "pv.2",
    eval      = "no"
    )
  str["volname"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Physical Volume Size",
    parameter = "",
    value     = "0",
    eval      = "no"
    )
  str["volsize"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Physical Volume Configuration",
    parameter = "part",
    value     = get_kickstart_volpart(str), 
    eval      = "get_kickstart_volpart(str)"
    )
  str["volpart"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Volume Group Name",
    parameter = "",
    value     = "VolGroup00",
    eval      = "no"
    )
  str["volgroupname"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Physical Extent Size",
    parameter = "",
    value     = "32768",
    eval      = "no"
    )
  str["pesize"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Volume Group Configuration",
    parameter = "volgroup",
    value     = get_kickstart_volgroup(str), 
    eval      = "get_kickstart_volgroup(str)"
    )
  str["volgroup"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Minimum Swap Size",
    parameter = "",
    value     = "512",
    eval      = "no"
    )
  str["swapmin"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Maximum Swap Size",
    parameter = "",
    value     = "1024",
    eval      = "no"
    )
  str["swapmax"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Swap Volume Name",
    parameter = "",
    value     = "LogVol01",
    eval      = "no"
    )
  str["swapvol"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Swap Logical Volume Configuration",
    parameter = "logvol",
    value     = get_kickstart_logswap(str), 
    eval      = "get_kickstart_logswap(str)"
    )
  str["logswap"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Root Filesystem",
    parameter = "",
    value     = "ext3",
    eval      = "no"
    )
  str["rootfs"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Root Volume Name",
    parameter = "",
    value     = "LogVol00",
    eval      = "no"
    )
  str["rootvol"]=cfg
  cfg=Cfg.new(
    type      = "",
    question  = "Root Size",
    parameter = "",
    value     = "1024",
    eval      = "no"
    )
  str["rootsize"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Root Logical Volume Configuration",
    parameter = "logvol",
    value     = get_kickstart_logroot(str), 
    eval      = "get_kickstart_logroot(str)"
    )
  str["logroot"]=cfg
  cfg=Cfg.new(
    type      = "output",
    question  = "Finish Command",
    parameter = "",
    value     = "reboot",
    eval      = "no"
    )
  str["finish"]=cfg
  return str
end

def output_kickstart_header(str,output_file)
  if output_file
    file=File.open(output_file, 'a')
  end 
  str.each do |key, value|
    if str[key].type == "output"
      if str[key].parameter == ""
        if output_file
          output=str[key].value+"\n"
          file.write(output)
        else
          puts str[key].value
        end
      else
        if output_file
          output=str[key].parameter+" "+str[key].value+"\n"
          file.write(output)
        else
          puts str[key].parameter+" "+str[key].value
        end
      end
    end
  end
  if output_file
    file.close
  end
  return
end

def verify_kickstart_header(str)
  puts
  puts "You will be presented with a set of questions followed by the default output"
  puts "If you are happy with the default output simply hit enter"
  puts
  change=0
  str.each do |key, value|
    if change == 1
      if str[key].eval != "no"
        new_value=str[key].eval
        new_value=eval"[#{new_value}]"
        str[key].value=new_value
      end
    end
    puts str[key].question+"?"
    puts "[ "+str[key].value+" ]"
    print "  "
    answer=gets.chomp
    if answer != ""
      if answer != str[key].value
        change=1
        str[key].value=answer 
      end
    end
  end
  return
end

def verify_kickstart_array(array)
  puts
  puts "You will be presented with the existing entries"
  puts
  array.each_with_index do |line, index|
    puts line
    puts "[ y/Y/n/N/yes|no ]"
    print "  "
    answer=gets.chomp
    if answer == "n" or answer == "N" or answer == "no"
      array[index]="#"+array[index]
    end
    if answer == "" or answer == "y" or answer == "Y" or answer == "yes"
      array[index]=answer 
    end
  end
  return array
end

def populate_kickstart_packages(pkg)
  pkg.push("@ core")
  pkg.push("grub")
  pkg.push("e2fsprogs")
  pkg.push("lvm2")
  pkg.push("kernel-devel")
  pkg.push("kernel-headers")
  return pkg
end

def output_kickstart_packages(pkg,output_file)
  if output_file
    file=File.open(output_file, 'a')
    output="\n%packages\n"
    file.write(output)
  else
    puts
    puts "%packages"
  end
  pkg.each do |pkg_name|
    if output_file
      output=pkg_name+"\n"
      file.write(output)
    else  
      puts pkg_name
    end
  end
  if output_file
    file.close
  end
  return
end

def populate_kickstart_post_vagrant_commands(cmd)
  cmd.push("/usr/sbin/groupadd vagrant")
  cmd.push("/usr/sbin/useradd vagrant -g vagrant -G wheel")
  cmd.push("echo \"vagrant\"|passwd --stdin vagrant")
  cmd.push("echo \"vagrant        ALL=(ALL)       NOPASSWD: ALL\" >> /etc/sudoers")
  return cmd
end

def populate_kickstart_post(cmd)
  cmd=populate_kickstart_post_vagrant_commands(cmd)
  return cmd
end

def output_kickstart_post(cmd,output_file)
  if output_file
    file=File.open(output_file, 'a')
    output="\n%post\n"
    file.write(output)
  else
    puts
    puts "%post"
  end
  cmd.each do |command|
    if output_file
      output=command+"\n"
      file.write(output)
    else
      puts command 
    end
  end
  if output_file
    file.close
  end
  return
end

def do_kickstart(config_name,output_file,output_type,base_dir)
  if output_file
    output_file="ks.cfg"
    if config_name
      output_dir=base_dir+"/"+config_name
      output_file=output_dir+"/"+output_file
      if !Dir.exists?(output_dir)
        Dir.mkdir(output_dir)
      end
    end
    if File.exists?(output_file)
      File.delete(output_file)
    end
  end
  str=Hash.new
  pkg=[]
  cmd=[]
  str=populate_kickstart_header(str)
  if output_type != "default"
    str=verify_kickstart_header(str)
  end
  pkg=populate_kickstart_packages(pkg)
  if output_type != "default"
    pkg=verify_kickstart_array(pkg)
  end
  cmd=populate_kickstart_post(cmd)
  if output_type != "default"
    cmd=verify_kickstart_array(cmd)
  end
  output_kickstart_header(str,output_file)
  output_kickstart_packages(pkg,output_file)
  output_kickstart_post(cmd,output_file)
  if output_file
    FileUtils.chmod(0755,output_file)
  end 
  return
end
